# DA Mini Project
* 주제 : 온라인 학습 플랫폼 사용자 행동 패턴 분석
* 개요 :
```
이번 프로젝트는 학습 서비스를 이용하는 독자들 중 누가 서비스 구독을 연장할지 예측하기 위한 사전 파악에 초점을 맞추고 있습니다. 우리의 목표는 유저들의 학습 습관, 선호도, 이용 행태 등 다양한 측면을 면밀히 분석하는 것입니다.


이를 위해, 우리는 유저들의 학습 패턴, 서비스 이용 시간, 콘텐츠에 대한 반응, 그리고 이들이 얼마나 자주 서비스를 이용하는지 등 다양한 데이터를 수집하고 분석할 계획입니다. 이 데이터는 유저들이 서비스에 얼마나 만족하고 있는지, 어떤 요소들이 그들의 구독 결정에 영향을 미치는지 이해하는 데 중요한 역할을 할 것입니다.
```
* 데이터셋 : mini.csv
    * 메타데이터
        * user_id: 사용자의 고유 식별자
        * subscription_duration: 사용자가 서비스에 가입한 기간 (월)
        * recent_login_time: 사용자가 마지막으로 로그인한 시간 (일)
        * average_login_time:  사용자의 일반적인 로그인 시간
        * average_time_per_learning_session: 각 학습 세션에 소요된 평균 시간 (분)
        * monthly_active_learning_days: 월간 활동적인 학습 일수
        * total_completed_courses: 완료한 총 코스 수
        * recent_learning_achievement: 최근 학습 성취도
        * abandoned_learning_sessions: 중단된 학습 세션 수
        * community_engagement_level: 커뮤니티 참여도
        * preferred_difficulty_level: 선호하는 난이도
        * subscription_type: 구독 유형
        * customer_inquiry_history: 고객 문의 이력
        * payment_pattern : 사용자의 지난 3개월 간의 결제 패턴을 10진수로 표현한 값.
          - 7: 3개월 모두 결제함
          - 6: 첫 2개월은 결제했으나 마지막 달에는 결제하지 않음
          - 5: 첫 달과 마지막 달에 결제함
          - 4: 첫 달에만 결제함
          - 3: 마지막 2개월에 결제함
          - 2: 가운데 달에만 결제함
          - 1: 마지막 달에만 결제함
          - 0: 3개월 동안 결제하지 않음
        * target: 사용자가 다음 달에도 구독을 계속할지 (1) 또는 취소할지 (0)를 나타냄

--- 

### 키워드
* AARRR
* 서비스 구독을 연장한 경험이 잇는 사용자 세그먼트는 어떤 특징이 있는가? -> 상관분석(X)
* 서비스 구독을 연장하지 않은 사용자 세그먼트는 어떤 특징이 있는가? -> 상관분석(X)
   
* payment_pattern에 따른 target 분류(X)
   
다음 달에도 구독을 계속할 사용자 세그먼트(1)와 다음 달에 구독을 하지 않을 사용자 세그먼트(0)와의 행동 패턴 차이

다음 달에도 구독을 계속하는데 영향을 주는 요소

  
어떤 파생변수를 만들 수 있을까?

### 질의점 
* 각 가격정책은 무엇을 의미하는가?  
   -> 가격정책은 Basic과 Premium 존재

### 알게 된 점
* payment_pattern이 0이어도 subscription_type이 Basic, Premium인 사람이 있음, subcription_type에는 Null값이 없음 -> 결제와 구독은 관계가 없다

### 계획
1. target에 따른 data describe

--- 

In [8]:
import pandas as pd
data = pd.read_table('data/mini.csv',sep=',')
data

user_id  subscription_duration  recent_login_time  average_login_time  \
0     b919c29d                     13                 14           14.946163   
1     a0a60abb                     16                 18           18.453224   
2     b9f171ae                     22                  1           16.195228   
3     5dc0ba8b                      1                 19           17.628656   
4     65c83654                      4                  5           21.390656   
...        ...                    ...                ...                 ...   
9995  ae6b76bc                     22                 29           14.727623   
9996  24588752                     10                 11           19.374054   
9997  e4622a54                      7                 27           18.240978   
9998  e07fbad9                     11                  7           18.783800   
9999  e12dcb55                     10                  5           13.073230   

      average_time_per_learning_session  monthly_active_learning_days  \
0                              8.427187                            18   
1                             72.646087                            16   
2                             21.774492                            13   
3                             42.659066                            19   
4                             30.744287                            19   
...                                 ...                           ...   
9995                          84.053558                            18   
9996                          45.464833                             9   
9997                         127.302411                            24   
9998                           5.297234                            10   
9999                          28.120031                             3   

      total_completed_courses  recent_learning_achievement  \
0                          16                    68.360455   
1                          13                    97.567322   
2                          14                    94.358763   
3                          18                    70.153228   
4                          10                    81.917908   
...                       ...                          ...   
9995                       16                    64.966803   
9996                        8                    82.750244   
9997                       14                    81.567839   
9998                       10                    89.885656   
9999                       13                    64.811297   

      abandoned_learning_sessions  community_engagement_level  \
0                               3                           4   
1                               2                           3   
2                               3                           4   
3                               0                           3   
4                               2                           4   
...                           ...                         ...   
9995                            2                           5   
9996                            3                           3   
9997                            3                           5   
9998                            4                           5   
9999                            4                           5   

     preferred_difficulty_level subscription_type  customer_inquiry_history  \
0                           Low             Basic                         4   
1                        Medium             Basic                         1   
2                        Medium           Premium                         0   
3                           Low             Basic                         1   
4                        Medium             Basic                         3   
...                         ...               ...                       ...   
9995                        Low           Premium                         1   
9996       

In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   user_id                            10000 non-null  object 
 1   subscription_duration              10000 non-null  int64  
 2   recent_login_time                  10000 non-null  int64  
 3   average_login_time                 10000 non-null  float64
 4   average_time_per_learning_session  10000 non-null  float64
 5   monthly_active_learning_days       10000 non-null  int64  
 6   total_completed_courses            10000 non-null  int64  
 7   recent_learning_achievement        10000 non-null  float64
 8   abandoned_learning_sessions        10000 non-null  int64  
 9   community_engagement_level         10000 non-null  int64  
 10  preferred_difficulty_level         10000 non-null  object 
 11  subscription_type                  10000 non-null  obje

In [81]:
data.describe().T

count       mean        std        min  \
subscription_duration              10000.0  11.897400   6.600896   1.000000   
recent_login_time                  10000.0  15.013200   8.362573   1.000000   
average_login_time                 10000.0  14.994076   3.001869   2.366189   
average_time_per_learning_session  10000.0  54.917720  56.024310   0.011515   
monthly_active_learning_days       10000.0  12.545400   6.932239   1.000000   
total_completed_courses            10000.0  12.227500   3.634125   1.000000   
recent_learning_achievement        10000.0  75.029513   9.968529  35.941755   
abandoned_learning_sessions        10000.0   3.043600   1.755052   0.000000   
community_engagement_level         10000.0   3.886100   1.262175   1.000000   
customer_inquiry_history           10000.0   2.010500   1.420983   0.000000   
payment_pattern                    10000.0   3.502900   2.311261   0.000000   
target                             10000.0   0.619900   0.485435   0.000000   

                                         25%        50%        75%         max  
subscription_duration               6.000000  12.000000  18.000000   23.000000  
recent_login_time                   8.000000  15.000000  22.000000   29.000000  
average_login_time                 13.025597  14.979228  16.995340   26.998490  
average_time_per_learning_session  15.276611  37.578818  75.584200  503.372616  
monthly_active_learning_days        7.000000  13.000000  19.000000   24.000000  
total_completed_courses            10.000000  12.000000  15.000000   27.000000  
recent_learning_achievement        68.278054  75.126061  81.718976  112.643828  
abandoned_learning_sessions         2.000000   3.000000   4.000000   12.000000  
community_engagement_level          3.000000   4.000000   5.000000    5.000000  
customer_inquiry_history            1.000000   2.000000   3.000000   10.000000  
payment_pattern                     1.000000   4.000000   6.000000    7.000000  
target                              0.000000   1.000000   1.000000    1.000000

In [38]:
data.groupby(['payment_pattern', 'target'])['user_id'].count().sort_index(ascending=False)

payment_pattern  target
7                1         795
                 0         489
6                1         801
                 0         467
5                1         750
                 0         488
4                1         753
                 0         461
3                1         763
                 0         467
2                1         747
                 0         459
1                1         783
                 0         502
0                1         807
                 0         468
Name: user_id, dtype: int64

In [40]:
data.groupby(['target'])['user_id'].count().sort_index(ascending=False)

target
1    6199
0    3801
Name: user_id, dtype: int64

___

In [53]:
data[data.target == 1]

user_id  subscription_duration  recent_login_time  average_login_time  \
1     a0a60abb                     16                 18           18.453224   
2     b9f171ae                     22                  1           16.195228   
3     5dc0ba8b                      1                 19           17.628656   
4     65c83654                      4                  5           21.390656   
6     682ca511                      8                 23           10.629897   
...        ...                    ...                ...                 ...   
9994  9ee4e40c                     11                  6           11.187836   
9995  ae6b76bc                     22                 29           14.727623   
9996  24588752                     10                 11           19.374054   
9997  e4622a54                      7                 27           18.240978   
9998  e07fbad9                     11                  7           18.783800   

      average_time_per_learning_session  monthly_active_learning_days  \
1                             72.646087                            16   
2                             21.774492                            13   
3                             42.659066                            19   
4                             30.744287                            19   
6                            116.455949                            12   
...                                 ...                           ...   
9994                          88.757610                             6   
9995                          84.053558                            18   
9996                          45.464833                             9   
9997                         127.302411                            24   
9998                           5.297234                            10   

      total_completed_courses  recent_learning_achievement  \
1                          13                    97.567322   
2                          14                    94.358763   
3                          18                    70.153228   
4                          10                    81.917908   
6                          10                    58.708390   
...                       ...                          ...   
9994                        9                    77.335911   
9995                       16                    64.966803   
9996                        8                    82.750244   
9997                       14                    81.567839   
9998                       10                    89.885656   

      abandoned_learning_sessions  community_engagement_level  \
1                               2                           3   
2                               3                           4   
3                               0                           3   
4                               2                           4   
6                               0                           2   
...                           ...                         ...   
9994                            5                           4   
9995                            2                           5   
9996                            3                           3   
9997                            3                           5   
9998                            4                           5   

     preferred_difficulty_level subscription_type  customer_inquiry_history  \
1                        Medium             Basic                         1   
2                        Medium           Premium                         0   
3                           Low             Basic                         1   
4                        Medium             Basic                         3   
6                           Low             Basic                         0   
...                         ...               ...                       ...   
9994                       High             Basic                         1   
9995       

In [52]:
data[data.target == 1].info()

<class 'pandas.core.frame.DataFrame'>
Index: 6199 entries, 1 to 9998
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   user_id                            6199 non-null   object 
 1   subscription_duration              6199 non-null   int64  
 2   recent_login_time                  6199 non-null   int64  
 3   average_login_time                 6199 non-null   float64
 4   average_time_per_learning_session  6199 non-null   float64
 5   monthly_active_learning_days       6199 non-null   int64  
 6   total_completed_courses            6199 non-null   int64  
 7   recent_learning_achievement        6199 non-null   float64
 8   abandoned_learning_sessions        6199 non-null   int64  
 9   community_engagement_level         6199 non-null   int64  
 10  preferred_difficulty_level         6199 non-null   object 
 11  subscription_type                  6199 non-null   object 
 1

In [82]:
data[data.target == 1].describe().T

count       mean        std        min  \
subscription_duration              6199.0  11.862720   6.590500   1.000000   
recent_login_time                  6199.0  14.985804   8.322310   1.000000   
average_login_time                 6199.0  14.974035   3.029275   2.366189   
average_time_per_learning_session  6199.0  60.105533  63.477523   0.013535   
monthly_active_learning_days       6199.0  12.571866   6.980077   1.000000   
total_completed_courses            6199.0  12.216003   3.636558   2.000000   
recent_learning_achievement        6199.0  75.047419  10.027058  36.115562   
abandoned_learning_sessions        6199.0   3.016616   1.754416   0.000000   
community_engagement_level         6199.0   3.895628   1.264061   1.000000   
customer_inquiry_history           6199.0   2.010808   1.428306   0.000000   
payment_pattern                    6199.0   3.500403   2.316925   0.000000   
target                             6199.0   1.000000   0.000000   1.000000   

                                         25%        50%        75%         max  
subscription_duration               6.000000  12.000000  18.000000   23.000000  
recent_login_time                   8.000000  15.000000  22.000000   29.000000  
average_login_time                 12.976618  14.979217  16.964874   26.284396  
average_time_per_learning_session  15.908562  39.705116  80.159683  503.372616  
monthly_active_learning_days        7.000000  12.000000  19.000000   24.000000  
total_completed_courses            10.000000  12.000000  15.000000   27.000000  
recent_learning_achievement        68.249866  75.199591  81.867694  112.643828  
abandoned_learning_sessions         2.000000   3.000000   4.000000   12.000000  
community_engagement_level          3.000000   4.000000   5.000000    5.000000  
customer_inquiry_history            1.000000   2.000000   3.000000    8.000000  
payment_pattern                     1.000000   3.000000   6.000000    7.000000  
target                              1.000000   1.000000   1.000000    1.000000

In [83]:
data[data.target == 0].describe().T

count       mean        std        min  \
subscription_duration              3801.0  11.953959   6.618295   1.000000   
recent_login_time                  3801.0  15.057880   8.428730   1.000000   
average_login_time                 3801.0  15.026762   2.956732   3.636424   
average_time_per_learning_session  3801.0  46.456986  39.639733   0.011515   
monthly_active_learning_days       3801.0  12.502236   6.854200   1.000000   
total_completed_courses            3801.0  12.246251   3.630554   1.000000   
recent_learning_achievement        3801.0  75.000310   9.873574  35.941755   
abandoned_learning_sessions        3801.0   3.087609   1.755430   0.000000   
community_engagement_level         3801.0   3.870560   1.259105   1.000000   
customer_inquiry_history           3801.0   2.009997   1.409145   0.000000   
payment_pattern                    3801.0   3.506972   2.302291   0.000000   
target                             3801.0   0.000000   0.000000   0.000000   

                                         25%        50%        75%         max  
subscription_duration               6.000000  12.000000  18.000000   23.000000  
recent_login_time                   8.000000  15.000000  22.000000   29.000000  
average_login_time                 13.077376  14.979239  17.030607   26.998490  
average_time_per_learning_session  14.485582  34.773346  68.298309  179.176060  
monthly_active_learning_days        7.000000  13.000000  18.000000   24.000000  
total_completed_courses            10.000000  12.000000  15.000000   27.000000  
recent_learning_achievement        68.316764  74.981448  81.520614  107.135153  
abandoned_learning_sessions         2.000000   3.000000   4.000000   11.000000  
community_engagement_level          3.000000   4.000000   5.000000    5.000000  
customer_inquiry_history            1.000000   2.000000   3.000000   10.000000  
payment_pattern                     1.000000   4.000000   6.000000    7.000000  
target                              0.000000   0.000000   0.000000    0.000000

In [64]:
pd.DataFrame(data[data.target == 1].describe().T['mean'])

mean
subscription_duration              11.862720
recent_login_time                  14.985804
average_login_time                 14.974035
average_time_per_learning_session  60.105533
monthly_active_learning_days       12.571866
total_completed_courses            12.216003
recent_learning_achievement        75.047419
abandoned_learning_sessions         3.016616
community_engagement_level          3.895628
customer_inquiry_history            2.010808
payment_pattern                     3.500403
target                              1.000000

In [66]:
pd.DataFrame(data[data.target == 0].describe().T['mean'])

mean
subscription_duration              11.953959
recent_login_time                  15.057880
average_login_time                 15.026762
average_time_per_learning_session  46.456986
monthly_active_learning_days       12.502236
total_completed_courses            12.246251
recent_learning_achievement        75.000310
abandoned_learning_sessions         3.087609
community_engagement_level          3.870560
customer_inquiry_history            2.009997
payment_pattern                     3.506972
target                              0.000000

### 수치형 데이터로 된 행동 데이터를 비교해본 결과, average_time_per_learning_session가 유의미한 차이가 보임

target == 1은 60.105533, target == 0은 46.456986

In [74]:
data[data.target == 1].groupby('preferred_difficulty_level')['user_id'].count()

preferred_difficulty_level
High      1241
Low       3047
Medium    1911
Name: user_id, dtype: int64

In [71]:
data[data.target == 0].groupby('preferred_difficulty_level')['user_id'].count()

preferred_difficulty_level
High       759
Low       1945
Medium    1097
Name: user_id, dtype: int64

### 선호 난이도 데이터를 비교분석한 결과, preferred_difficulty_level이 다름 

target 1이 0에 비해서 Medium, High 비율이 높음

In [78]:
data[data.target == 1].groupby('subscription_type')['user_id'].count()

subscription_type
Basic      3683
Premium    2516
Name: user_id, dtype: int64

In [79]:
data[data.target == 0].groupby('subscription_type')['user_id'].count()

subscription_type
Basic      2296
Premium    1505
Name: user_id, dtype: int64

### 선호 난이도 데이터를 비교분석한 결과, preferred_difficulty_level이 다름 

target 1이 0에 비해서 Premium 비율이 높음